### PR TITLE
Add support for culling

### PR DESCRIPTION
**Pull request by andrewbelu**
Adds support for culling. Introduces a new mod aggregator to get the max value of a mod.

With no culling effect active:
image
image
image

With 10% cull active:
image
image

Breakdown:
image

### CODE REVIEW

**Comment by Nostrademous** [comment]
Looks good and cleanly implemented. I do wonder if the Max aggregator is needed though as I believe (without looking into it) it could be implemented using Tabulate and taking the max value when iterating the resulting table. If it’s more performant than sure, but curious.

**Comment by andrewbelu** [comment]
> Looks good and cleanly implemented. I do wonder if the Max aggregator is needed though as I believe (without looking into it) it could be implemented using Tabulate and taking the max value when iterating the resulting table. If it’s more performant than sure, but curious.

I just hadn't really taken a good look at it (been looking at the code base for about a day now). I've refactored (and tested) it to just re-use the Tabulate functionality.

File: Classes/ModList.lua [hunk 1]
```diff
@@ -139,7 +139,6 @@ function ModListClass:ListInternal(context, result, cfg, flags, keywordFlags, so
 		for i = 1, #self do
 			local mod = self[i]
 			if mod.name == modName and mod.type == "LIST" and band(flags, mod.flags) == mod.flags and MatchKeywordFlags(keywordFlags, mod.keywordFlags) and (not source or mod.source:match("[^:]+") == source) then
-				local value
```
Comment (ppoelzl) [inline Classes/ModList.lua:142]: Is this a fix related to culling? If not, please create new PRs for unrelated changes.

File: Modules/CalcOffence.lua [hunk 2]
```diff
@@ -1735,6 +1735,16 @@ function calcs.offence(env, actor, activeSkill)
...
 		output.ScaledDamageEffect = output.ScaledDamageEffect * output.DoubleDamageEffect
 		
+		-- Calculate culling DPS
+		local criticalCull = skillModList:Max(cfg, "CriticalCullPercent") or 0
+		if criticalCull > 0 then
+			criticalCull = criticalCull * (output.CritChance / 100)
+		end
+		local regularCull = skillModList:Max(cfg, "CullPercent") or 0;
+		local maxCullPercent = m_max(criticalCull, regularCull);
+		globalOutput.CullPercent = maxCullPercent;
```
Comment (ppoelzl) [inline Modules/CalcOffence.lua:1745]:
```suggestion
		local regularCull = skillModList:Max(cfg, "CullPercent") or 0
		local maxCullPercent = m_max(criticalCull, regularCull)
		globalOutput.CullPercent = maxCullPercent
```
Style: only use semicolons for multiple statements on the same line, and also don't do multiple statements on the same line.

File: Modules/CalcOffence.lua [hunk 3]
```diff
@@ -1735,6 +1735,16 @@ function calcs.offence(env, actor, activeSkill)
...
+		globalOutput.CullMultiplier = 100 / (100 - globalOutput.CullPercent)
+
```
Comment (ppoelzl) [inline Modules/CalcOffence.lua:1747]:
```suggestion
		globalOutput.CullMultiplier = 100 / (100 - globalOutput.CullPercent)
```
Style: avoid excessive empty lines.

File: Modules/CalcOffence.lua [hunk 4]
```diff
@@ -2082,7 +2092,9 @@ function calcs.offence(env, actor, activeSkill)
 		-- Calculate average damage and final DPS
 		output.AverageHit = totalHitAvg * (1 - output.CritChance / 100) + totalCritAvg * output.CritChance / 100
 		output.AverageDamage = output.AverageHit * output.HitChance / 100
-		output.TotalDPS = output.AverageDamage * (globalOutput.HitSpeed or globalOutput.Speed) * (skillData.dpsMultiplier or 1)
+
+		output.TotalDPS = output.AverageDamage * (globalOutput.HitSpeed or globalOutput.Speed) * (skillData.dpsMultiplier or 1) * globalOutput.CullMultiplier
+
```
Comment (ppoelzl) [inline Modules/CalcOffence.lua:2097]:
```suggestion
		output.TotalDPS = output.AverageDamage * (globalOutput.HitSpeed or globalOutput.Speed) * (skillData.dpsMultiplier or 1) * globalOutput.CullMultiplier
```

File: Modules/CalcOffence.lua [hunk 5]
```diff
@@ -2167,6 +2179,9 @@ function calcs.offence(env, actor, activeSkill)
 		if skillData.dpsMultiplier then
 			t_insert(breakdown.TotalDPS, s_format("x %g ^8(DPS multiplier for this skill)", skillData.dpsMultiplier))
 		end
+		if output.CullMultiplier and output.CullMultiplier > 1.0 then
```
Comment (ppoelzl) [inline Modules/CalcOffence.lua:2182]:
```suggestion
		if output.CullMultiplier and output.CullMultiplier > 1 then
```
No need to use floating-point numbers here. In Lua < 5.3, every number is a double.

File: Modules/ModParser.lua [hunk 6]
```diff
@@ -2373,6 +2374,15 @@ local specialModList = {
...
 	["knockback direction is reversed"] = { mod("EnemyKnockbackDistance", "MORE", -200) },
+	-- Culling
+	["culling strike"] = { mod("CullPercent", "MAX", 10)},
+	["hits with this weapon have culling strike against bleeding enemies"] = { mod("CullPercent", "MAX", 10, { type = "ActorCondition", actor = "enemy", var = "Bleeding" })},
+	["you have culling strike against cursed enemies"] = { mod("CullPercent", "MAX", 10, { type = "ActorCondition", actor = "enemy", var = "Cursed" })},
+	["critical strikes have culling strike"] = { mod("CriticalCullPercent", "MAX", 10 )},
+	["your spells have culling strike"] = { mod("CullPercent", "MAX", 10, nil, ModFlag.Spell )},
+	["culling strike against burning enemies"] = { mod("CullPercent", "MAX", 10, { type = "ActorCondition", actor = "enemy", var = "Burning" })},
+	["culling strike against marked enemy"] = { mod("CullPercent", "MAX", 10, { type = "ActorCondition", actor = "enemy", var = "Marked" })},
+
```
Comment (ppoelzl) [inline Modules/ModParser.lua:2385]:
```suggestion
	["culling strike"] = { mod("CullPercent", "MAX", 10) },
	["hits with this weapon have culling strike against bleeding enemies"] = { mod("CullPercent", "MAX", 10, { type = "ActorCondition", actor = "enemy", var = "Bleeding" }) },
	["you have culling strike against cursed enemies"] = { mod("CullPercent", "MAX", 10, { type = "ActorCondition", actor = "enemy", var = "Cursed" }) },
	["critical strikes have culling strike"] = { mod("CriticalCullPercent", "MAX", 10) },
	["your spells have culling strike"] = { mod("CullPercent", "MAX", 10, nil, ModFlag.Spell) },
	["culling strike against burning enemies"] = { mod("CullPercent", "MAX", 10, { type = "ActorCondition", actor = "enemy", var = "Burning" }) },
	["culling strike against marked enemy"] = { mod("CullPercent", "MAX", 10, { type = "ActorCondition", actor = "enemy", var = "Marked" }) },
```
Style: balanced whitespace around brackets.

**Review by ppoelzl — CHANGES_REQUESTED** [review]

File: Classes/ModList.lua (hunk 1 above)
Comment (andrewbelu) [reply Classes/ModList.lua:142]: It is just a cleanup of a useless variable I saw, not related to the change. I can revert it if you'd like.

**Comment by andrewbelu** [comment]
Note, I've updated the CalcOffence so that the culling multiplier is applied to the combined output, rather than the total DPS, just so any ailment/impale builds that have cull will be accounted for.

File: Classes/ModList.lua (hunk 1 above)
Comment (ppoelzl) [reply Classes/ModList.lua:142]: Please revert this change then.

File: Data/SkillStatMap.lua [hunk 7]
```diff
@@ -1513,5 +1513,9 @@ return {
 },
 ["support_added_cooldown_count_if_not_instant"] = {
 	mod("AdditionalCooldownUses", "BASE", nil)
+},
+["kill_enemy_on_hit_if_under_10%_life"] = {
+	mod("CullPercent", "MAX", nil), 
+	value = 10
 }
```
Comment (ppoelzl) [inline Data/SkillStatMap.lua:1520]:
```suggestion
},
```
Please add a trailing comma for smaller Git diffs in future PRs here.

**Review by ppoelzl — APPROVED** [review]
LGTM now.
